pandas/core/frame.py

```diff
@@ -5039,8 +5039,12 @@ def append(self, other, ignore_index=False, verify_integrity=False):
                 # index name will be reset
                 index = Index([other.name], name=self.index.name)
 
-            combined_columns = self.columns.tolist() + self.columns.union(
-                other.index).difference(self.columns).tolist()
+            idx_diff = other.index.difference(self.columns)
+            try:
+                combined_columns = self.columns.append(idx_diff)
+            except TypeError:
```

Comment:
so this should not be done here, rather *if* a TypeError happens, then ``Index.append`` handles.

Reply:
I don't understand you here, could you expand.

Reply:
```
In [1]: pd.Index([1, 2, 3]).append(pd.Index(list('abc')))
Out[1]: Index([1, 2, 3, 'a', 'b', 'c'], dtype='object')

In [2]: pd.Index([1, 2, 3]).append(pd.CategoricalIndex(list('abc')))
Out[2]: Index([1, 2, 3, 'a', 'b', 'c'], dtype='object')

In [3]: pd.Index([1, 2, 3]).append(pd.MultiIndex.from_product([[1], ['a']]))
Out[3]: Index([1, 2, 3, (1, 'a')], dtype='object')

In [4]: pd.Index([1, 2, 3]).append(pd.IntervalIndex.from_breaks([1, 2, 3]))
Out[4]: Index([1, 2, 3, (1, 2], (2, 3]], dtype='object')

```

.append`` should *never* raise

so you are catching something that does not happen.

Reply:
The below fail though:

```python
>>> pd.CategoricalIndex('A B C'.split()).append(pd.Index([1])
TypeError: cannot append a non-category item to a CategoricalIndex
```

This is what is tested in the ``if not isinstance(df_columns, (pd.IntervalIndex, pd.MultiIndex)):`` part of the tests. In addition, the reason why pd.IntervalIndex and pd.MultiIndex are excluded for testing is because of reindexing issues:

```python
>>> ii = pd.IntervalIndex.from_breaks([0,1,2])
>>> df = pd.DataFrame([[1, 2], [4, 5]], columns=ii)
>>> ser = pd.Series([7], index=['a'])
>>> combined = pd.Index(ii.tolist() + ser.index.tolist())
>>> df.reindex(columns=combined)
TypeError: unorderable types: Interval() > int()
```

Reply:
hmm, the MI case we did on purpose. However you cannot convert to lists and expect things to work. for the TypeError, do ``.astype(object)`` on each left and right then ``.append`` again. we *never* should coerce to lists.